text/0061-error-format-and-definitions.md

```diff
@@ -0,0 +1,810 @@
+- Title: Error Format and Definitions
+- Start Date: 2021-08-13
+- Specification PR: [#61](https://github.com/meilisearch/specifications/pull/61)
+- Discovery Issue: [#44](https://github.com/meilisearch/product/issues/44#issuecomment-896121211)
+
+# Error Format
+
+## 1. Functional Specification
+
+### I. Summary
+
+The error format is updated to no longer contain unnecessary `error` prefix terms in the names of these attributes. e.g. `errorType`. The context is clear enough to understand that this is an `error` object.
+
+This specification also serves as a reference point for the complete list of API errors that the user may encounter.
+
+#### Summary Key Points
+
+- `error` is removed from the attributes name and `type`values.
+- An exhaustive list of API errors is defined. See Errors List part.
+
+### II. Motivation
+
+The main motivation is to stabilize the current `error` resource to a version that conforms to our API convention and thus allows future evolutions on a more solid base. This specification avoids adding unnecessary information in the error object's attribute names.
```

Comment:
> that conforms to our API convention and thus allows future evolutions on a more solid base

Why is the mention of the word `error` something we want to remove? It seems a maybe useful  information for those checking errors using conditions from the response body instead of catching errors through HTTP code. 

The latter is of course the best practice but I was wondering why we decided to remove the word.

[Algolia](https://www.algolia.com/doc/rest-api/search/#error-handling) does not mention the word error

ElasticSearch does
exampl: 
```json
{
    "error": {
        "caused_by": {
            "reason": "json_parse_exception: Unexpected character (',' (code 44)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name\n at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@139163f; line: 3, column: 17]",
            "type": "json_parse_exception"
        },
        "reason": "failed to parse",
        "root_cause": [
            {
                "reason": "json_parse_exception: Unexpected character (',' (code 44)): was expecting either valid name character (for unquoted name) or double-quote (for quoted) to start field name\n at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@139163f; line: 3, column: 17]",
                "type": "json_parse_exception"
            }
        ],
        "type": "mapper_parsing_exception"
    },
    "status": 400
}
```

Maybe `message` could be `error` instead, or wrap everything is an error key? 
```json
{
   "error": { "message": "woops", ... }
}
```

I did not find a consensus online on best practices so the way you propose it might be the best! I have no strong opinions on it :)

Reply:
I think there is no need to specify `errorCode` or `errorType` rather than `code` or `type` since the user understands that he is dealing with an error by reading the content of the response. The programs themselves should be based on the HTTP code. In any case, the context is clear so that there is no redundancy of `error` in the names.

Concerning the asynchronous errors that are reported directly in the task object, the error object will be displayed under an `error` key.

Reply:
> Concerning the asynchronous errors that are reported directly in the task object, the error object will be displayed under an error key.

Should it be stated in this spec?

It would look like this? 

```json
{
  "status": "failed",
  "updateId": 1,
  "type": {
    "name": "DocumentsAddition",
  },
  "enqueuedAt": "2019-12-07T21:10:07.607581330Z",
  "duration": 0.000048524,
  "processedAt": "2019-12-07T21:10:20.511525620Z",
  "error": {
    "message": "You must have an authorization token",
    "code": "missing_authorization_header",
    "type": "authentication_error",
    "link": "https://docs.meilisearch.com/errors#missing_authorization_header"
  }
}
 ```


> Your last proposal is also something we can do! I've come across a few examples of APIs doing this because errors were returned with a 200 HTTP code response while some also return the status code in the error object with an equivalent HTTP code for the response. 

This might be linked indeed to how errors might pop in the API! I'm not sure this is something that might happen in MeiliSearch. 
It could also introduce the `warn` key, but it might be something that is never returned in API responses. At least i don't remember crossing one